numba/cuda/tests/cudapy/test_dispatcher.py

```diff
@@ -272,6 +272,78 @@ def pi_sin_array(x, n):
         self.assertIsInstance(regs_per_thread, int)
         self.assertGreater(regs_per_thread, 0)
 
+    def test_get_shared_mem_per_block_unspecialized(self):
+        # A kernel where the register usage per thread is likely to differ
+        # between different specializations
+        @cuda.jit
+        def simple_smem_float(ary):
+            sm = cuda.shared.array(100, dtype=float32)
+            i = cuda.grid(1)
+            if i == 0:
+                for j in range(100):
+                    sm[j] = j
+            cuda.syncthreads()
+            ary[i] = sm[i]
+
+        @cuda.jit
+        def simple_smem_double(ary):
+            sm = cuda.shared.array(100, dtype=float64)
+            i = cuda.grid(1)
+            if i == 0:
+                for j in range(100):
+                    sm[j] = j
+            cuda.syncthreads()
+            ary[i] = sm[i]
+
+        # Call the kernel with different arguments to create two different
+        # definitions within the Dispatcher object
+        N = 100
+        arr_f32 = np.zeros(N, dtype=np.float32)
+        arr_f64 = np.zeros(N, dtype=np.float64)
+
+        simple_smem_float[1, N](arr_f32)
+        simple_smem_double[1, N](arr_f64)
+
+        sig_f32 = void(float32[::1])
+        sig_f64 = void(float64[::1])
+        shared_mem_per_block_f32 = \
+            simple_smem_float.get_shared_mem_per_block(sig_f32)
+        shared_mem_per_block_f64 = \
+            simple_smem_double.get_shared_mem_per_block(sig_f64)
+        self.assertIsInstance(shared_mem_per_block_f32, int)
+        self.assertIsInstance(shared_mem_per_block_f64, int)
+
+        self.assertEqual(shared_mem_per_block_f32, 400)
+        self.assertEqual(shared_mem_per_block_f64, 800)
+
+        # Check that getting the shared memory per blockfor all signatures
+        # provides the same values as getting the shared mem per block for
+        # individual signatures.
+        shared_mem_per_block_f32_all = \
+            simple_smem_float.get_shared_mem_per_block()
+        shared_mem_per_block_f64_all = \
+            simple_smem_double.get_shared_mem_per_block()
+        self.assertEqual(shared_mem_per_block_f32_all[sig_f32.args],
+                         shared_mem_per_block_f32)
+        self.assertEqual(shared_mem_per_block_f64_all[sig_f64.args],
+                         shared_mem_per_block_f64)
+
+    def test_get_shared_mem_per_block_specialized(self):
+        @cuda.jit(void(float32[::1]))
+        def simple_smem(ary):
+            sm = cuda.shared.array(100, dtype=float32)
+            i = cuda.grid(1)
+            if i == 0:
+                for j in range(100):
+                    sm[j] = j
+            cuda.syncthreads()
+            ary[i] = sm[i]
+
+        # Check we get a positive integer for the specialized variation
```

Comment:
I think we know what shared memory usage to expect (unlike the register usage test above) so we should be able to assert that we have exactly the shared memory usage we expect.